### PR TITLE
Notes

### PR DESCRIPTION
**Pull request by h16nning**
> [!WARNING]
> These changes are breaking. The database will need to act as expected. I didn't bother implement db upgrade logic because it would have been very time-consuming and without real users... :/
>
> Checking this pr out locally may also make it difficult to run lower db versions again.

### What are notes?
Notes replace shared values. They store:
- an unique id
- creationDate
- customOrder to be ordered in notebook view
- reference to a deck
- content

Notes share a lot of the behavior of shared values (refactored most of the code). For example, they get deleted as there are no more cards referencing to them.

#### Content of notes
- card type specific
- all card types store their content now in a note (even normal cards which wouldn't need content shared over multiple cards)

#### What is still stored in cards?
- content, but only that specific to a single card (e. g. occlusionNumber in cloze cards)
- model (and history): review informationen (every card of a note, e. g. different cloze fields might be learned at different speeds)
- id, type, reference to deck and its note
- customOrder, creationDate and preview: might remove those as it is pretty much redundant information that could also be drawn from the cards note. Still need to decide.

### Advantages of notes
- Listing notes will not create "duplicate entries" #65.
- It would be possible to create notes that have now cards attached. This could be useful to add plain notes to NotebookView that shall not appear when you review your cards.

### Where to use notes and where cards?
- notes: spotlight search, notebook → wherever to look at cards without reviewing them
- cards: learn view
- card manager: maybe both (see below)

### Disadvantage of notes
- By creating notes, a similar level of complexity to anki will be introduced (as discussed in #61). It is important to make the difference between cards and notes logical. Is a card manager where you can toggle searching between cards and notes useful?
- An additional layer of redundancy between notes, decks and cards is introduced → more efforts to be made for consistent data.

### What doesn't work yet / any more?
- There is "Note Menu" yet. In notebook view it is now missing (previously there was a card menu)
- Moving cards will note move the note. Should that even happen? It is possible to have cards of a note spread across multiple decks. The note will stay where it was created.
- Option to toggle between show answer in notebook view is not working yet. More complex logic needs to implement. For example: Toggling cloze fields individually by clicking on them.
- I haven't checked all features yet. Bugs are to be expected...

I re-used a lot of old code. There is still a lot of cleaning up to be done...

### CODE REVIEW

**Comment by eMerzh** [comment]
Hey, very nice you've start a move on this :) ... I hadn't had a lot of time to look at this in details yet … 

if I understand correctly Notes are the "values" you want to learn & all, while cards stores more or less the history of revisions.

If it's correct, i would completely hide the "card" from the users (even though notes will then probably be called card in the UI 🤪 ). Notes/cards relation should stay an implementation detail (no toggle or something ... at least by default).

I would even go one step further in moving the table as "history" or "stats" or something similar that referenced a the value with maybe some kind of "key" you could use as an index for cloze, or double sided etc... .
but i guess, it's what cards are 🤔  (i can see preview around [here](https://github.com/h16nning/skola/pull/73/files#diff-aea4b87592ca726cd3762c89096e9c970d287ebaef3bd8d3265f5db6290301e3L39), i guess it's the old code?)

anyway, cool to see this happening, well done :) 
i'll try to check in more details later

**Comment by h16nning** [comment]
> if I understand correctly Notes are the "values" you want to learn & all, while cards stores more or less the history of revisions.
I'd say so, cards as an atomic piece of *learnable* information (one cloze field, one image occlusion field, a question → answer card etc.)

Notes as something that tie it all together.

#### Hiding implementation details
I agree with hiding most of the implementation details...😅 However, I see some use cases where it would be useful to see all cards in a list. For example, I like to be able to see my Anki cards sorted by difficulty, interval, etc:

Bildschirmfoto 2024-02-25 um 23 02 39

The question is if integrating this into a combined card / note manager like Anki has is a smart idea. As I am writing about it, I see the following solution. Let me know what you think about that:

1. Make the current manager view to be notes only. Make it editing, moving notes around and deleting focussed. Editing information as notes is less confusing than the current way (editing a card and it "silently" updates linked cards in the background).

2. Have a separate, new view focussed for cards. Here the focus lies on seeing the actual cards. Options to sort them by their performance (difficulty, relapses, ...), reschedule and disable them etc. Maybe it could also be a little bit like a statistics view.



> I would even go one step further in moving the table as "history" or "stats" or something similar that referenced a the value with maybe some kind of "key" you could use as an index for cloze, or double sided etc... .

I'm sorry, I don't understand what you mean by that 😅

File: src/logic/note.ts [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+  customOrder?: number;
+}
+
+export interface Note<T extends CardType> extends NoteSkeleton {
+  content: NoteContent<T>;
+}
+
+export function createNoteSkeleton(deck: string): NoteSkeleton {
+  return {
+    deck: deck,
```
Comment (eMerzh) [inline src/logic/note.ts:22]: not really sure why you keep the deck ref in here...
I guess it's for easier access on learning phase or something? 

but you need to "join" to cards anyways no ?

File: src/logic/note.ts [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    creationDate: new Date(Date.now()),
+    id: uuidv4(),
+    referencedBy: [],
```
Comment (eMerzh) [inline src/logic/note.ts:25]: wouldn't be "easier" to just fetch cards using the "note" field? with a query?

File: src/logic/note.ts (hunk 1 above)
Comment (h16nning) [reply src/logic/note.ts:22]: I found it pretty handy to access it directly in several cases, like in NoteMenu, everything with searching and filtering.

Actually, I would rather consider removing the deck property from cards... 😅

File: src/logic/note.ts (hunk 2 above)
Comment (h16nning) [reply src/logic/note.ts]: Thanks!

implemented it, should work

I think there might be some other things that could be simplified by taking more advantage of Dexies's features...

**Pull request merged**